### PR TITLE
Add pre commit "bugbear" checks

### PR DESCRIPTION
**Pull request by benjeffery**
[flake8-bugbear](https://pypi.org/project/flake8-bugbear/) checks for likely bugs and design problems. It found a few which are fixed in this PR.

### CODE REVIEW

File: docs/development.rst [hunk 1]
```diff
@@ -33,11 +33,19 @@ Quickstart
...
+  Sometimes these will report "files were modified by this hook" ``git add``
+  and ``git commit --amend`` will update the commit with the automatically modified
+  version.
+  The modifications made are for consistency, code readability and designed to
+  minimise merge conflicts. They are guaranteed not to modify the functionality of the
+  code. To run the checks without committing use ``pre-commit run``. To bypass
+  the checks (to save or get feedback on work-in-progress) use ``git commit
+  --no-verify``
+- When ready open a pull request on GitHub. Please make sure that the tests pass before
+  you open the PR.
```
Comment (petrelharp) [inline docs/development.rst]: Don't we want people to be able to ask for help in getting tests to pass as part of the PR?

File: python/tests/test_lowlevel.py [hunk 2]
```diff
@@ -1832,7 +1838,7 @@ def f(mutations):
         u = ts.get_num_nodes()
         for bad_node in [u, u + 1, 2 * u]:
             self.assertRaises(_tskit.LibraryError, f, [(0.1, bad_node)])
-        for bad_pos in [-1, length, length + 1]:
+        for _bad_pos in [-1, length, length + 1]:
```
Comment (petrelharp) [inline python/tests/test_lowlevel.py]: Error here, length should be bad_pos

File: python/tests/test_tree_stats.py [hunk 3]
```diff
@@ -2894,33 +2894,32 @@ def naive_branch_allele_frequency_spectrum(
     for j in range(num_windows):
         begin = windows[j]
         end = windows[j + 1]
-        for set_index, sample_set in enumerate(sample_sets):
```
Comment (petrelharp) [inline python/tests/test_tree_stats.py]: Yep, don't need this.

**Comment by jeromekelleher** [comment]
Just needs a rebase to update after #475 was merged (ditto for the other PRs) so we can see the diff.

Thanks for all this @benjeffery, it's going to be major upgrade in terms of our dev processes!

File: docs/development.rst (hunk 1 above)
Comment (benjeffery) [reply docs/development.rst]: I've added an extra note to that effect.

File: python/tests/test_lowlevel.py (hunk 2 above)
Comment (benjeffery) [reply python/tests/test_lowlevel.py]: Fixed in https://github.com/tskit-dev/tskit/pull/479/commits/0900c826fb1c821ef35e87784bae4f94b895f8f9

**Comment by jeromekelleher** [comment]
LGTM, happy to merge whenever.

**Comment by jeromekelleher** [comment]
Note: ready to merge after 478 goes in.

**Comment by jeromekelleher** [comment]
I think this is next in line @benjeffery? Do you want to squash the commits, or will I use the "squash and merge" button? (I don't think this allows me to edit the commit messages though)

**Comment by benjeffery** [comment]
Squashed!

**Comment by jeromekelleher** [comment]
Great, thanks, merging. (I can't be bothered waiting for CI here, it's already passed a bunch of times).

**Pull request merged**